### PR TITLE
[improve] [client]Add new ServiceUrlProvider implementation: SameAuthParamsAutoClusterFailover

### PR DESCRIPTION
**Pull request by poorbarcode**
### Motivation

**Background**
So far, there are two implementations of `ServiceUrlProvider`
- `ControlledClusterFailover`: call an HTTP API to change the pulsar service URL.
- `AutoClusterFailover`: check the pulsar service URL by creating a socket, and automatically switch to the high-priority pulsar service URL.

**Issue**
- We build two clusters `c1` and `c2`, and start a binary-way replication between them. 
- Enabled `AutoClusterFailover` for our client to switch connections to the healthy cluster when the primary one can not work.
- The network traffic is:`client -> proxy -> brokers`

If all the brokers crash, the proxies are still healthy, `AutoClusterFailover` can not determine that the primary cluster can not provide service anymore.

### Modifications

Add a new ServiceUrlProvider implementation: `SameAuthParamsAutoClusterFailover`, which uses `lookup topic` to check whether the target cluster is healthy or not. Since the clusters that enabled Geo-Replication always have the same auth settings, it can be easy to implement(skipping the different auth settings ).

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: x

### CODE REVIEW

**Review by congbobo184 — APPROVED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/NonDurableSubscriptionTest.java [hunk 1]
```diff
@@ -542,18 +545,34 @@ public void testReaderInitAtDeletedPosition() throws Exception {
...
-            assertTrue(actMarkDeletedPos.compareTo(expectedMarkDeletedPos) >= 0);
+            AssertJUnit.assertTrue(actMarkDeletedPos.compareTo(expectedMarkDeletedPos) >= 0);
         });
 
+        admin.topics().createSubscription(topicName, "s2", MessageId.earliest);
+        admin.topics().createSubscription(topicName, "s3", MessageId.latest);
+        PersistentTopicInternalStats internalStats = admin.topics().getInternalStats(topicName);
+        ManagedLedgerInternalStats.CursorStats cursorStats2 = internalStats.cursors.get("s2");
+        String[] ledgerIdAndEntryId2 = cursorStats2.markDeletePosition.split(":");
+        ImmutablePositionImpl actMarkDeletedPos2 =
```
Comment (lhotari) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/api/NonDurableSubscriptionTest.java:566]: There shouldn't be a need to use ImmutablePositionImpl in code. Simply use Position as the type and PositionFactory to create instances. They will be immutable by default.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/NonDurableSubscriptionTest.java [hunk 2]
```diff
@@ -542,18 +545,34 @@ public void testReaderInitAtDeletedPosition() throws Exception {
...
-                    PositionFactory.create(msgIdInDeletedLedger5.getLedgerId(), msgIdInDeletedLedger5.getEntryId());
+            ImmutablePositionImpl actMarkDeletedPos =
+                    new ImmutablePositionImpl(Long.valueOf(ledgerIdAndEntryId[0]), Long.valueOf(ledgerIdAndEntryId[1]));
+            ImmutablePositionImpl expectedMarkDeletedPos =
+                    new ImmutablePositionImpl(msgIdInDeletedLedger5.getLedgerId(), msgIdInDeletedLedger5.getEntryId());
+            log.info("LAC: {}", internalStats.lastConfirmedEntry);
             log.info("Expected mark deleted position: {}", expectedMarkDeletedPos);
             log.info("Actual mark deleted position: {}", cursorStats.markDeletePosition);
-            assertTrue(actMarkDeletedPos.compareTo(expectedMarkDeletedPos) >= 0);
+            AssertJUnit.assertTrue(actMarkDeletedPos.compareTo(expectedMarkDeletedPos) >= 0);
```
Comment (lhotari) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/api/NonDurableSubscriptionTest.java:558]: AssertJUnit shouldn't be needed

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/NonDurableSubscriptionTest.java
Comment (lhotari) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/api/NonDurableSubscriptionTest.java:1]: What is the purpose of this test in this PR? Is it unrelated?

Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/api/NonDurableSubscriptionTest.java:1]: Created a new PR to revert it
- https://github.com/apache/pulsar/pull/23168

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/NonDurableSubscriptionTest.java (hunk 2 above)
Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/api/NonDurableSubscriptionTest.java:558]: Created a new PR to revert it
- https://github.com/apache/pulsar/pull/23168

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/NonDurableSubscriptionTest.java (hunk 1 above)
Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/api/NonDurableSubscriptionTest.java:566]: Created a new PR to revert it
- https://github.com/apache/pulsar/pull/23168

**Pull request merged**